### PR TITLE
Fix EEXIST crash when reusing pdfToPngConvertOptions.outputFolder across comparePdf calls

### PR DESCRIPTION
**Pull request by dichovsky**
## Summary

- Fixed `comparePdf` / `comparePdfDetailed` crashing with `ComparePdfRenderingError` (EEXIST) on the **second** invocation when `pdfToPngConvertOptions.outputFolder` is reused, while the first call succeeds.
- Added regression coverage for running `comparePdf` twice against the same `outputFolder`.

## Root Cause

`comparePdf` renders pages to disk through `pdf-to-png-converter`, whose filesystem sink opens each PNG with an exclusive-create flag (`'wx'`) and **refuses to overwrite** an existing file. Its own `savePNGfile` docstring instructs callers to "clear the output folder before re-running the same conversion if they expect to reuse the same output names."

`src/internal/renderPdfPages.ts` (`renderPdfPage`) never performed that clearing step. On the first run the namespace files (`<outputFolder>/actual/<name>`, `<outputFolder>/expected/<name>`) are created; on any subsequent run with the same folder the renderer hits `EEXIST`, which `runPdfToPng` wraps into the generic `ComparePdfRenderingError: Failed to render <side> PDF pages.` — hiding the real cause from the caller.

Reproduction (before fix):
```
Run 1: true
Run 2: THREW ComparePdfRenderingError - Failed to render actual PDF pages.
       cause: EEXIST ... /test-results/.../pages/actual/comparePdf_1.png
```

## Fix

`renderPdfPage` now removes the exact PNG file it is about to regenerate, immediately before rendering, when an `outputFolder` is configured:

```ts
rmSync(join(outputFolder, sourceLabel, outputFileMaskFunc(pageNumber)), { force: true });
```

This is the minimal correct fix:
- It is **surgical** — it removes only the single file the library owns and is re-creating (`outputFolder` is library-resolved, the `actual/`/`expected/` namespaces are library-created, and `outputFileMaskFunc` always has a library default), never any unrelated file.
- `force: true` makes the first run a no-op (file absent) and removes a stale leaf — including a planted leaf symlink, which `rmSync` unlinks **without following**, so no redirect is honored. The folder/ancestor symlink defenses in `renderOutputFolderGuards` are untouched.
- Public API and types are unchanged.

## Tests

Run in the worktree (`node_modules` symlinked from main checkout):

- `npx vitest run --coverage` → **165 passed (30 files)**, coverage **100% statements / 100% branches / 100% functions / 100% lines** (thresholds: 100/90/100/100).
- New regression test `should re-render to a reused outputFolder without an exclusive-create EEXIST failure`:
  - **Before fix:** fails with `EEXIST` on the second `comparePdf` call (verified by reverting only the source change).
  - **After fix:** passes; both runs return `true` and the rendered PNGs exist.
- `npm run lint`, `npm run codemap:check`, `npm run build`, `npm run test:types`, `npm run test:artifacts` → all pass. `CODEMAP.md` regenerated (`npm run codemap`) to track the new internal helper.

### Pre-existing, unrelated failure
`npm run test:license` fails on **both** this branch and untouched `main` because the transitive dev dependency `lightningcss-darwin-arm64@1.32.0` is MPL-2.0. It is environmental and unrelated to this change (which touches no dependencies).

## Risk

Low. The change is confined to `src/internal/renderPdfPages.ts` and adds one regression test.
- Only affects the file-backed render path (`outputFolder` set); the default in-memory path is unchanged (`clearStaleRenderTarget` early-returns).
- No public API, type, or behavioral change for first-run / single-run callers.
- Affected area: per-page render output file lifecycle.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by dichovsky — COMMENTED** [review]
## Code Review — fix/reused-output-folder-eexist

**Verdict: APPROVE (comment-only — reviewer is the PR author).** All verification checks pass. Evidence below.

---

### 1. Bug is REAL — CONFIRMED

I independently verified the bug by reverting `src/internal/renderPdfPages.ts` to the `main` version and running the regression test:

```
FAIL  __tests__/8.compare.with.custom.pdf.options.test.ts
  > should re-render to a reused outputFolder without an exclusive-create EEXIST failure

  ComparePdfRenderingError: Failed to render actual PDF pages.
  Caused by: Error: EEXIST: file already exists, open '.../pages/actual/rerun_1.png'
    at savePNGfile node_modules/pdf-to-png-converter/out/outputWriter.js:39:16
```

`outputWriter.js` line 39 is exactly `const fd = await node_fs_1.promises.open(resolvedFilePath, 'wx')`. The `'wx'` flag is the direct cause — it opens with O_EXCL and refuses to overwrite. The library's own JSDoc says "Callers should clear the output folder before re-running the same conversion if they expect to reuse the same output names." This PR implements that documented contract.

---

### 2. Root-cause analysis is CORRECT

`pdf-to-png-converter/out/outputWriter.js:39` uses `open(path, 'wx')` (exclusive-create). On a second `comparePdf` call with the same `outputFolder`, the previous run's PNGs already exist, causing EEXIST. The library owns no cleanup path. Root cause confirmed.

---

### 3. Fix is minimal and targeted — CONFIRMED

`git diff main...HEAD -- src/types/ src/index.ts` produces no output — zero public API or type surface changes. The fix adds exactly one private function (`clearStaleRenderTarget`) and one call site inside `renderPdfPage`. No adjacent code was touched. No unrelated changes found.

---

### 4. Regression test FAILS before fix / PASSES after — CONFIRMED

- Without fix: 1 test failed (EEXIST), 3 passed
- With fix: all 4 tests pass

---

### 5. Security — PASS

**`rmSync` with `force: true` removes only the leaf symlink, not its target.** Verified empirically: planted a symlink at `render/actual/test_1.png` pointing to an external file `outside/secret.txt`, ran `rmSync` on the symlink path, confirmed the symlink was removed while `secret.txt` remained intact. Node.js `rmSync` with `force: true` does not follow symlinks at the leaf.

**Path construction is bounded.** The path is `join(outputFolder, sourceLabel, outputFileMaskFunc(pageNumber))` where:
- `outputFolder` has already been resolved and validated by `validateRenderOutputFolder` (symlink-walks the full ancestor chain, then pre-creates and lstat-asserts each `actual/`/`expected/` namespace as a real directory)
- `sourceLabel` is a literal `'actual' | 'expected'` from the type system
- `outputFileMaskFunc(pageNumber)` is a caller-supplied filename

**`renderOutputFolderGuards.ts` is untouched** — confirmed by diff. The CWE-59/CWE-61 guards and TOCTOU defense via pre-creation + lstat remain intact.

---

### 6. No regressions — CONFIRMED

Full suite with coverage:
```
Test Files  30 passed (30)
Tests      165 passed (165)

Statements : 100% (318/318)
Branches   : 100% (252/252)
Functions  :  100% (64/64)
Lines      :  100% (314/314)
```

---

### 7. Edge cases — all handled

- **`outputFolder` set, `outputFileMaskFunc` unset:** `clearStaleRenderTarget` returns early. Not reachable in practice because `normalizeComparisonOptions` (line 28-30) unconditionally injects a default `outputFileMaskFunc` before any rendering begins.
- **In-memory render (no `outputFolder`):** early return — no filesystem access.
- **Multi-page docs:** `renderPdfPage` is called once per page sequentially; `clearStaleRenderTarget` clears exactly the one file about to be re-rendered.
- **Metadata-only listing path (`listRenderedPageNumbers`):** uses `returnMetadataOnly: true`; `pdf-to-png-converter` skips `savePNGfile` entirely in that mode. No EEXIST possible; `clearStaleRenderTarget` is not on this path.

---

### Pre-existing issue (unrelated to this PR)

`npm run test:license` fails identically on `main` due to transitive dev dependency `lightningcss-darwin-arm64@1.32.0` (MPL-2.0). Not introduced by this PR.

File: src/internal/renderPdfPages.ts [hunk 1]
```diff
@@ -64,6 +67,28 @@ export async function renderPdfPage(
...
+    pageNumber: number,
+    sourceLabel: 'actual' | 'expected',
+): void {
+    const { outputFolder, outputFileMaskFunc } = pdfToPngConvertOpts;
+    if (!outputFolder || !outputFileMaskFunc) {
+        return;
+    }
+
+    rmSync(join(outputFolder, sourceLabel, outputFileMaskFunc(pageNumber)), { force: true });
+}
```
[the remark this answers is not in the document]
Comment (dichovsky) [reply src/internal/renderPdfPages.ts:123]: Addressed in 1345c4c. `clearStaleRenderTarget` now (1) resolves the target and removes it only when `isPathWithinRoot(targetPath, <outputFolder>/<sourceLabel>)` holds — a mask that escapes the namespace is left untouched so the renderer's own `savePNGfile` containment check stays the single source of truth (no out-of-namespace unlink possible); and (2) wraps `rmSync` failures (EISDIR/EACCES/etc.) in `ComparePdfConfigurationError` with the original `cause`, so no raw ErrnoException leaks out of `comparePdfDetailed`. Added unit coverage for both branches in `17.internal.render-pdf-pages.test.ts`.

File: src/internal/renderPdfPages.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import { rmSync } from 'node:fs';
 import { join } from 'node:path';
 import { pdfToPng } from 'pdf-to-png-converter';
 import { ComparePdfRenderingError } from '../errors/ComparePdfRenderingError.js';
```
[the remark this answers is not in the document]
Comment (dichovsky) [reply src/internal/renderPdfPages.ts:5]: Done in 1345c4c — `ComparePdfConfigurationError` is now imported and thrown by `clearStaleRenderTarget` when `rmSync` fails, keeping comparePdf's typed error contract intact.

File: src/internal/renderPdfPages.ts [hunk 3]
```diff
@@ -64,6 +69,48 @@ export async function renderPdfPage(
...
+    const { outputFolder, outputFileMaskFunc } = pdfToPngConvertOpts;
+    if (!outputFolder || !outputFileMaskFunc) {
+        return;
+    }
+
+    const renderNamespace = resolve(outputFolder, sourceLabel);
+    const targetPath = resolve(renderNamespace, outputFileMaskFunc(pageNumber));
+    if (!isPathWithinRoot(targetPath, renderNamespace)) {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (dichovsky) [reply src/internal/renderPdfPages.ts:113]: Addressed in 74a75b8. `clearStaleRenderTarget` now evaluates `outputFileMaskFunc(pageNumber)` inside a guarded block: a mask function that throws is wrapped in `ComparePdfConfigurationError` (preserving the original `cause`), mirroring the existing `rmSync` failure handling — so no raw error escapes `comparePdf` / `comparePdfDetailed`. The renderer is never reached when the mask throws. Added unit coverage (`17.internal.render-pdf-pages.test.ts`) asserting the typed error, the preserved cause, and that `pdfToPng` is not invoked. Full suite: 168 passed, 100% statements/branches/functions/lines.
